src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1628 @@
+  // the flow analysis. This method returns the index of a location, allocating
+  // one if this is the first time we see it.
+  LocationIndex getIndex(const Location& location) {
+    auto iter = locationIndexes.find(location);
+    if (iter != locationIndexes.end()) {
+      return iter->second;
+    }
+
+    // Allocate a new index here.
+    size_t index = locations.size();
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+    std::cout << "  new index " << index << " for ";
+    dump(location);
+#endif
+    if (index >= std::numeric_limits<LocationIndex>::max()) {
+      // 32 bits should be enough since each location takes at least one byte
+      // in the binary, and we don't have 4GB wasm binaries yet... do we?
+      Fatal() << "Too many locations for 32 bits";
+    }
+    locations.emplace_back(location);
+    locationIndexes[location] = index;
+
+    return index;
+  }
+
+  IndexLink getIndexes(const LocationLink& link) {
+    return {getIndex(link.from), getIndex(link.to)};
+  }
+
+  // See the comment on CollectedFuncInfo::childParents. This is the merged info
+  // from all the functions and the global scope.
+  std::unordered_map<LocationIndex, LocationIndex> childParents;
+
+  // The work remaining to do during the flow: locations that we need to flow
+  // content from, after new content reached them.
+  //
+  // Using a set here is efficient as multiple updates may arrive to a location
+  // before we get to processing it.
+  //
+  // The items here could be {location, newContents}, but it is more efficient
+  // to have already written the new contents to the main data structure. That
+  // avoids larger data here, and also, updating the contents as early as
+  // possible is helpful as anything reading them meanwhile (before we get to
+  // their work item in the queue) will see the newer value, possibly avoiding
+  // flowing an old value that would later be overwritten.
+#ifdef POSSIBLE_CONTENTS_INSERT_ORDERED
+  InsertOrderedSet<LocationIndex> workQueue;
+#else
+  std::unordered_set<LocationIndex> workQueue;
+#endif
+
+  // Maps a heap type + an index in the type (0 for an array) to the index of a
+  // SpecialLocation for a cone read of those contents. We use such special
+  // locations because a read of a cone type (as opposed to an exact type) will
+  // require N incoming links, from each of the N subtypes - and we need that
+  // for each struct.get of a cone. If there are M such gets then we have N * M
+  // edges for this. Instead, make a single canonical "cone read" location, and
+  // add a single link to it from each get, which is only N + M (plus the cost
+  // of adding "latency" in requiring an additional step along the way for the
+  // data to flow along).
+  std::unordered_map<std::pair<HeapType, Index>, LocationIndex>
+    canonicalConeReads;
+
+  // Creates a new special location that is different from all others so far.
+  LocationIndex makeSpecialLocation() {
```

Comment:
Maybe `makeFreshLocation` to be more descriptive?

Reply:
This name is confusing to me... Other locations are not unique? How do you define uniqueness in location sense? For example, are `ExpressionLocation`s not unique?

I prefer `SpecialLocation`, because it is... well, special, but if we want to emphasize that this function makes a new unique location each time it is called, how about keeping `SpecialLocation` struct name as before and rename this function to `makeNewSpecialLocation` or `makeUniqueSpecialLocation` or something? Sorry for reopening the resolved discussions, and I don't feel strongly, but this name is little confusing to me.

Reply:
We talked briefly about this offline. Other names like `FreshLocation`, `ConeLocation`, and `AbstractLocation` were suggested. Among the candidates so far (`SpecialLocation`, `UniqueLocation`, `FreshLocation`, `AbstractLocation`, `ConeLocation`), I personally prefer one of `SpecialLocation`, `AbstractLocation`, or `ConeLocation`. I think `UniqueLocation` is confusing because it implies other locations are not unique and it's hard to define uniqueness in the location sense, and `FreshLocation` does not say much about what that is... Are other locations not fresh?

But anyway, it's my personal preferences, and if you too prefer something else, I'm fine with that too. And it's not that this name is super important...

Reply:
I also don't feel strongly. How about `ConeReadLocation`? (just to clarify it is just for reads and not writes)

Reply:
I like that too!

Reply:
Ok, refactored to `ConeReadLocation`. This does make the code simpler, to just focus on the concrete use case for now. Seems worth it, as I'm not sure if we'll need something more later that would benefit from a generic "unique/special" location.